presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java

```diff
@@ -2215,6 +2216,69 @@ public PhysicalOperation visitExchange(ExchangeNode node, LocalExecutionPlanCont
         {
             checkArgument(node.getScope() == LOCAL, "Only local exchanges are supported in the local planner");
 
+            if (node.getOrderingScheme().isPresent()) {
+                return createLocalMerge(node, context);
+            }
+            else {
+                return createLocalExchange(node, context);
+            }
+        }
+
+        private PhysicalOperation createLocalMerge(ExchangeNode node, LocalExecutionPlanContext context)
+        {
+            checkState(node.getSources().size() == 1, "single source is expected");
+
+            // local merge source must have a single driver
+            context.setDriverInstanceCount(1);
+
+            PlanNode sourceNode = getOnlyElement(node.getSources());
+            LocalExecutionPlanContext subContext = context.createSubContext();
+            PhysicalOperation source = sourceNode.accept(this, subContext);
+
+            int operatorsCount = subContext.getDriverInstanceCount().orElse(1);
+            List<Type> types = getSourceOperatorTypes(node, context.getTypes());
+            LocalExchangeFactory exchangeFactory = new LocalExchangeFactory(
+                    node.getPartitioningScheme().getPartitioning().getHandle(),
+                    operatorsCount,
+                    types,
+                    ImmutableList.of(),
+                    Optional.empty(),
+                    source.getPipelineExecutionStrategy(),
+                    maxLocalExchangeBufferSize);
+
+            List<OperatorFactory> operatorFactories = new ArrayList<>(source.getOperatorFactories());
+            List<Symbol> expectedLayout = node.getInputs().get(0);
+            Function<Page, Page> pagePreprocessor = enforceLayoutProcessor(expectedLayout, source.getLayout());
+            operatorFactories.add(new LocalExchangeSinkOperatorFactory(
+                    exchangeFactory,
+                    subContext.getNextOperatorId(),
+                    node.getId(),
+                    exchangeFactory.newSinkFactoryId(),
+                    pagePreprocessor));
+            context.addDriverFactory(subContext.isInputDriver(), false, operatorFactories, subContext.getDriverInstanceCount(), source.getPipelineExecutionStrategy());
+            // the main driver is not an input... the exchange sources are the input for the plan
+            context.setInputDriver(false);
+
+            OrderingScheme orderingScheme = node.getOrderingScheme().get();
+            List<Integer> sortChannels = orderingScheme.getOrderBy().stream()
+                    .map(argument -> node.getOutputSymbols().indexOf(argument))
+                    .collect(toImmutableList());
+            List<SortOrder> orderings = orderingScheme.getOrderBy().stream()
+                    .map(orderingScheme::getOrdering)
+                    .collect(toImmutableList());
```

Comment:
for consistency sake, consider using `makeLayout`, `getChannelsForSymbols` and `OrderingScheme#getOrderingList`:

```
ImmutableMap<Symbol, Integer> layout = makeLayout(node);
List<Integer> sortChannels = getChannelsForSymbols(orderingScheme.getOrderBy(), layout);
List<SortOrder> sortOrder = orderingScheme.getOrderingList();
```

and then reuse `layout` variable in a call to `new PhysicalOperation` on L2277.